File: engine/competition/competition-manager.ts
import type { Club, MatchConfig, MatchFixture } from "@/engine/models";
import { simulate } from "@/engine/simulation";
import { KnockoutBracket } from "./knockout-bracket";
import { MundialBracket } from "./mundial-bracket";

/**
 * Orchestrates the entire season: national knockouts per country + mundial.
 */
export class CompetitionManager {
  private nationalBracketsMap: Map<string, KnockoutBracket>;
  private _mundialBracket: MundialBracket | null = null;
  private readonly clubs: Club[];
  private readonly _baseSeed: number;

  get nationalBrackets(): ReadonlyMap<string, KnockoutBracket> {
    return this.nationalBracketsMap;
  }

  get mundialBracket(): MundialBracket | null {
    return this._mundialBracket;
  }

  get baseSeed(): number {
    return this._baseSeed;
  }

  constructor(clubs: Club[], baseSeed: number) {
    this.clubs = clubs;
    this._baseSeed = baseSeed;
    this.nationalBracketsMap = new Map<string, KnockoutBracket>();

    // Group clubs by country and create national brackets
    const byCountry = new Map<string, number[]>();
    for (const club of clubs) {
      const list = byCountry.get(club.country) ?? [];
      list.push(club.id);
      byCountry.set(club.country, list);
    }

    for (const [country, teamIds] of byCountry) {
      this.nationalBracketsMap.set(country, new KnockoutBracket(teamIds));
    }
  }

  /**
   * Restores a competition manager from saved state. Used by the persistence layer.
   */
  static restore(
    clubs: Club[],
    baseSeed: number,
    nationalBrackets: Map<string, KnockoutBracket>,
    mundialBracket: MundialBracket | null,
  ): CompetitionManager {
    const cm = new CompetitionManager(clubs, baseSeed);
    cm.nationalBracketsMap.clear();
    for (const [key, bracket] of nationalBrackets) {
      cm.nationalBracketsMap.set(key, bracket);
    }
    cm._mundialBracket = mundialBracket;
    return cm;
  }

  /**
   * Generates fixtures for the current national round across all countries.
   */
  generateNationalRound(matchDay: number): MatchFixture[] {
    const allFixtures: MatchFixture[] = [];
    for (const bracket of this.nationalBracketsMap.values()) {
      if (!bracket.isFinished) {
        allFixtures.push(...bracket.generateNextRound(matchDay));
      }
    }
    return allFixtures;
  }

  /**
   * Simulates all fixtures for a given day and records results.
   */
  simulateFixtures(
    fixtures: readonly MatchFixture[],
    daySeed: number,
  ): void {
    for (let i = 0; i < fixtures.length; i++) {
      const fixture = fixtures[i];
      const home = this.clubs.find((c) => c.id === fixture.homeClubId)!;
      const away = this.clubs.find((c) => c.id === fixture.awayClubId)!;

      const config: MatchConfig = {
        homeTeam: home.team,
        awayTeam: away.team,
        seed: daySeed + i,
        homeTactics: null,
        awayTactics: null,
        homeSubstitutions: null,
        awaySubstitutions: null,
        homeCard: null,
        awayCard: null,
      };

      const result = simulate(config);
      fixture.result = [
        result.finalState.scoreHome,
        result.finalState.scoreAway,
      ];
    }
  }

  /**
   * Advances all national brackets after a round is played.
   */
  advanceNationalRounds(): void {
    for (const bracket of this.nationalBracketsMap.values()) {
      if (!bracket.isFinished) {
        bracket.advanceRound();
      }
    }
  }

  /**
   * Returns true if all national brackets are finished.
   */
  areNationalsFinished(): boolean {
    for (const bracket of this.nationalBracketsMap.values()) {
      if (!bracket.isFinished) return false;
    }
    return true;
  }

  /**
   * Creates the mundial bracket from national champions.
   * Call only after all nationals are finished.
   */
  createMundial(): void {
    if (!this.areNationalsFinished()) {
      throw new Error(
        "Cannot create mundial before nationals finish.",
      );
    }

    const champions: number[] = [];
    for (const bracket of this.nationalBracketsMap.values()) {
      champions.push(bracket.championId!);
    }

    this._mundialBracket = new MundialBracket(champions);
  }

  /**
   * Generates fixtures for the current mundial round.
   */
  generateMundialRound(matchDay: number): MatchFixture[] {
    if (this._mundialBracket === null) {
      throw new Error("Mundial not yet created.");
    }
    return this._mundialBracket.generateNextRound(matchDay);
  }

  /**
   * Advances the mundial bracket after a round.
   */
  advanceMundialRound(): void {
    this._mundialBracket?.advanceRound();
  }

  /**
   * Gets the national champion of the country where the given club plays.
   */
  getNationalChampion(country: string): number | null {
    const bracket = this.nationalBracketsMap.get(country);
    return bracket?.championId ?? null;
  }

  /**
   * Gets the mundial champion, or null if not yet decided.
   */
  getMundialChampion(): number | null {
    return this._mundialBracket?.championId ?? null;
  }

  /**
   * Returns true if the entire season (nationals + mundial) is complete.
   */
  isSeasonComplete(): boolean {
    return (
      this.areNationalsFinished() &&
      (this._mundialBracket?.isFinished ?? false)
    );
  }
}

File: src/competition/CompetitionManager.cs
using ElevenLegends.Data.Enums;
using ElevenLegends.Data.Models;
using ElevenLegends.Simulation;

namespace ElevenLegends.Competition;

/// <summary>
/// Orchestrates the entire season: national knockouts per country + mundial.
/// </summary>
public sealed class CompetitionManager
{
    private readonly Dictionary<string, KnockoutBracket> _nationalBrackets;
    private MundialBracket? _mundialBracket;
    private readonly List<Club> _clubs;
    private readonly int _baseSeed;

    public IReadOnlyDictionary<string, KnockoutBracket> NationalBrackets =>
        _nationalBrackets;
    public MundialBracket? MundialBracket => _mundialBracket;

    public CompetitionManager(List<Club> clubs, int baseSeed)
    {
        _clubs = clubs;
        _baseSeed = baseSeed;
        _nationalBrackets = new Dictionary<string, KnockoutBracket>();

        // Group clubs by country and create national brackets
        var byCountry = clubs.GroupBy(c => c.Country);
        foreach (var group in byCountry)
        {
            var teamIds = group.Select(c => c.Id).ToList();
            _nationalBrackets[group.Key] = new KnockoutBracket(teamIds);
        }
    }

    /// <summary>
    /// Restores a competition manager from saved state. Used by the persistence layer.
    /// </summary>
    internal static CompetitionManager Restore(
        List<Club> clubs,
        int baseSeed,
        Dictionary<string, KnockoutBracket> nationalBrackets,
        MundialBracket? mundialBracket)
    {
        var cm = new CompetitionManager(clubs, baseSeed);
        cm._nationalBrackets.Clear();
        foreach (var kvp in nationalBrackets)
            cm._nationalBrackets[kvp.Key] = kvp.Value;
        cm._mundialBracket = mundialBracket;
        return cm;
    }

    /// <summary>
    /// Generates fixtures for the current national round across all countries.
    /// </summary>
    public List<MatchFixture> GenerateNationalRound(int matchDay)
    {
        var allFixtures = new List<MatchFixture>();
        foreach (var bracket in _nationalBrackets.Values)
        {
            if (!bracket.IsFinished)
                allFixtures.AddRange(bracket.GenerateNextRound(matchDay));
        }
        return allFixtures;
    }

    /// <summary>
    /// Simulates all fixtures for a given day and records results.
    /// </summary>
    public void SimulateFixtures(IReadOnlyList<MatchFixture> fixtures, int daySeed)
    {
        for (int i = 0; i < fixtures.Count; i++)
        {
            var fixture = fixtures[i];
            var home = _clubs.First(c => c.Id == fixture.HomeClubId);
            var away = _clubs.First(c => c.Id == fixture.AwayClubId);

            var config = new MatchConfig
            {
                HomeTeam = home.Team,
                AwayTeam = away.Team,
                Seed = daySeed + i
            };

            var result = MatchSimulator.Simulate(config);
            fixture.Result = (result.FinalState.ScoreHome, result.FinalState.ScoreAway);
        }
    }

    /// <summary>
    /// Advances all national brackets after a round is played.
    /// </summary>
    public void AdvanceNationalRounds()
    {
        foreach (var bracket in _nationalBrackets.Values)
        {
            if (!bracket.IsFinished)
                bracket.AdvanceRound();
        }
    }

    /// <summary>
    /// Returns true if all national brackets are finished.
    /// </summary>
    public bool AreNationalsFinished() =>
        _nationalBrackets.Values.All(b => b.IsFinished);

    /// <summary>
    /// Creates the mundial bracket from national champions.
    /// Call only after all nationals are finished.
    /// </summary>
    public void CreateMundial()
    {
        if (!AreNationalsFinished())
            throw new InvalidOperationException("Cannot create mundial before nationals finish.");

        var champions = _nationalBrackets.Values
            .Select(b => b.ChampionId!.Value)
            .ToList();

        _mundialBracket = new MundialBracket(champions);
    }

    /// <summary>
    /// Generates fixtures for the current mundial round.
    /// </summary>
    public List<MatchFixture> GenerateMundialRound(int matchDay)
    {
        if (_mundialBracket == null)
            throw new InvalidOperationException("Mundial not yet created.");

        return [.. _mundialBracket.GenerateNextRound(matchDay)];
    }

    /// <summary>
    /// Advances the mundial bracket after a round.
    /// </summary>
    public void AdvanceMundialRound()
    {
        _mundialBracket?.AdvanceRound();
    }

    /// <summary>
    /// Gets the nacional champion of the country where the given club plays.
    /// </summary>
    public int? GetNationalChampion(string country) =>
        _nationalBrackets.TryGetValue(country, out var bracket)
            ? bracket.ChampionId
            : null;

    /// <summary>
    /// Gets the mundial champion, or null if not yet decided.
    /// </summary>
    public int? GetMundialChampion() => _mundialBracket?.ChampionId;

    /// <summary>
    /// Returns true if the entire season (nationals + mundial) is complete.
    /// </summary>
    public bool IsSeasonComplete() =>
        AreNationalsFinished() && (_mundialBracket?.IsFinished ?? false);
}
